### ISSUE TITLE
Cartoon Network - AttributeError: 'NoneType' object has no attribute 'groups'

### ISSUE THREAD

**Issue by along517** [rg3/youtube-dl#17224]
## Please follow the guide below

- You will be asked some questions and requested to provide some information, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *issue* (like this: `[x]`)
- Use the *Preview* tab to see what your issue will actually look like

---

### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2018.08.04*. If it's not, read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x ] I've **verified** and **I assure** that I'm running youtube-dl **2018.08.04**

### Before submitting an *issue* make sure you have:
- [x ] At least skimmed through the [README](https://github.com/rg3/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [ x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones
- [x ] Checked that provided video/audio/playlist URLs (if any) are alive and playable in a browser

### What is the purpose of your *issue*?
- [x ] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### The following sections concretize particular purposed issues, you can erase any section (the contents between triple ---) not applicable to your *issue*

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

Add the `-v` flag to **your command line** you run youtube-dl with (`youtube-dl -v <your command line>`), copy the **whole** output and insert it here. It should look similar to one below (replace it with **your** log inserted between triple ```):

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-v', 'https://www.cartoonnetwork.com/video/teen-titans/divide-and-conquer-episode.html']
[debug] Encodings: locale cp1252, fs mbcs, out cp437, pref cp1252
[debug] youtube-dl version 2018.08.04
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.17134
[debug] exe versions: none
[debug] Proxy map: {}
[CartoonNetwork] divide-and-conquer: Downloading webpage
Traceback (most recent call last):
  File "__main__.py", line 19, in <module>
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpckoq891b\build\youtube_dl\__init__.py", line 472, in main
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpckoq891b\build\youtube_dl\__init__.py", line 462, in _real_main
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpckoq891b\build\youtube_dl\YoutubeDL.py", line 2001, in download
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpckoq891b\build\youtube_dl\YoutubeDL.py", line 792, in extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpckoq891b\build\youtube_dl\extractor\common.py", line 502, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpckoq891b\build\youtube_dl\extractor\cartoonnetwork.py", line 28, in _real_extract
AttributeError: 'NoneType' object has no attribute 'groups'

```

---

### If the purpose of this *issue* is a *site support request* please provide all kinds of example URLs support for which should be included (replace following example URLs by **yours**):
- Single video: https://www.cartoonnetwork.com/video/teen-titans/divide-and-conquer-episode.html

Note that **youtube-dl does not support sites dedicated to [copyright infringement](https://github.com/rg3/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free)**. In order for site support request to be accepted all provided example URLs should not violate any copyrights.

---

### Description of your *issue*, suggested solution and other information

Videos from Cartoon Network do not work. It says AttributeError: 'NoneType' object has no attribute 'groups'.

**Comment by jhotmann** [comment]
Running into this as well. May take a stab at creating a fix but I'll have to familiarize myself with how youtube-dl works.

**Comment by GoNotCougsWA** [comment]
I have the same problem too but I found a fix and created a Pull request on the issue but most of it isn't my code and was from a different pull request that wasn't accepted.

**Comment by along517** [comment]
I tried using your pull request fix, but a different problem arose. It said "ERROR: Missing Authentication token".   Is there a fix to that? Thanks.

**Comment by GoNotCougsWA** [comment]
Oh, that error message, it's not because of me. It's means that your trying to view something that needs you to login with your tv provider to view here is how to login: https://github.com/rg3/youtube-dl#adobe-pass-options. If you don't have a tv provider you can view/download free episode that are full unlocked or clips.

**Comment by along517** [comment]
Thanks for all the help. The Adobe pass has not been working for me either.

**Comment by along517** [comment]
It works! Hopefully, youtube-dl accepts the pull request.

**Comment by along517** [comment]
Sorry to bring this up again, but it seems as though youtube-dl only records up to the 10-minute mark (where the first advertisement shows up). Is there a way to get past the ad? Is this the result of the code or should I be doing something different? Thanks.

**Comment by JamesTheAwesomeDude** [comment]
Sadly, I don't think @TheDaChicken's patch has fixed it.

After installing pycrypto and ffmpeg, with ytdl latest (as of this comment) git master:

````
$ youtube-dl https://www.cartoonnetwork.com/video/unlocked/when-it-rains-episode.html  -v --hls-prefer-native
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['https://www.cartoonnetwork.com/video/unlocked/when-it-rains-episode.html', '-v', '--hls-prefer-native']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.09.01
[debug] Python version 3.5.2 (CPython) - Linux-4.10.0-38-generic-x86_64-with-Ubuntu-16.04-xenial
[debug] exe versions: ffmpeg 3.4.1-1, ffprobe 3.4.1-1
[debug] Proxy map: {}
[CartoonNetwork] when-it-rains: Downloading webpage
Traceback (most recent call last):
  File "/home/james/.local/bin/youtube-dl", line 9, in <module>
    load_entry_point('youtube-dl==2018.9.1', 'console_scripts', 'youtube-dl')()
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/__init__.py", line 472, in main
    _real_main(argv)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/__init__.py", line 462, in _real_main
    retcode = ydl.download(all_urls)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/YoutubeDL.py", line 2001, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/YoutubeDL.py", line 792, in extract_info
    ie_result = ie.extract(url)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/common.py", line 502, in extract
    ie_result = self._real_extract(url)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/cartoonnetwork.py", line 28, in _real_extract
    id_type, video_id = re.search(r"_cnglobal\.cvp(Video|Title)Id\s*=\s*'([^']+)';", webpage).groups()
AttributeError: 'NoneType' object has no attribute 'groups'
````

**Comment by TheDaChicken** [comment]
It seems that your still using the old version of the code. I don't have "id_type, video_id = re.search(r"_cnglobal\.cvp(Video|Title)Id\s*=\s*'([^']+)';", webpage).groups()" in my code.

**Comment by TheDaChicken** [comment]
It seems my code thinks you need to login to a TV Provider. This could be easily fixed. By changing "auth_required" to nothing, as I know of. I will fix this.

**Comment by JamesTheAwesomeDude** [comment]
>Make sure your using the latest version of my code, since I didn't run into this problem.

The above logs were generated by telling pip to pull STRAIGHT from `git+https://github.com/TheDaChicken/youtube-dl`, so it's definitely fresh from the tap. :(

>By changing "auth_required" to nothing, as I know of. I will fix this.

Awesome. Standing by for the push!

**Comment by TheDaChicken** [comment]
Ok! Try it again now, I think I fixed it since youtube-dl isn't telling me that is trying to signing into cartoonnetwork.

**Comment by JamesTheAwesomeDude** [comment]
Sadly, doesn't seem to have fixed:

````
$ youtube-dl https://www.cartoonnetwork.com/video/unlocked/when-it-rains-episode.html -v --hls-prefer-native)
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['https://www.cartoonnetwork.com/video/unlocked/when-it-rains-episode.html', '-v', '--hls-prefer-native']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.08.04
[debug] Python version 3.5.2 (CPython) - Linux-4.10.0-38-generic-x86_64-with-Ubuntu-16.04-xenial
[debug] exe versions: ffmpeg 3.4.1-1, ffprobe 3.4.1-1
[debug] Proxy map: {}
[CartoonNetwork] when-it-rains: Downloading webpage
[CartoonNetwork] 0e9df1345ed47078df3cf75bec73f74562a4fdb8: Downloading JSON metadata
ERROR: This video is only available for users of participating TV providers. Use --ap-mso to specify Adobe Pass Multiple-system operator Identifier and --ap-username and --ap-password or --netrc to provide account credentials.
Traceback (most recent call last):
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/YoutubeDL.py", line 792, in extract_info
    ie_result = ie.extract(url)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/common.py", line 502, in extract
    ie_result = self._real_extract(url)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/cartoonnetwork.py", line 46, in _real_extract
    'auth_required': auth_required,
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/turner.py", line 212, in _extract_ngtv_info
    m3u8_url, media_id, ap_data or {}, tokenizer_query)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/turner.py", line 40, in _add_akamai_spe_token
    query['accessToken'] = self._extract_mvpd_auth(ap_data['url'], content_id, ap_data['site_name'], ap_data['site_name'])
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/adobepass.py", line 1409, in _extract_mvpd_auth
    raise_mvpd_required()
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/adobepass.py", line 1374, in raise_mvpd_required
    'and --ap-username and --ap-password or --netrc to provide account credentials.', expected=True)
youtube_dl.utils.ExtractorError: This video is only available for users of participating TV providers. Use --ap-mso to specify Adobe Pass Multiple-system operator Identifier and --ap-username and --ap-password or --netrc to provide account credentials.
````

And I immediately double-checked to *ensure* that latest commit is in:
````
$ grep -C3 "auth_required = 'false'" ~/.local/lib/python3.5/site-packages/youtube_dl/extractor/cartoonnetwork.py 
        title = self._html_search_regex(r'[^>]+.episodeTitle = "(.+?)"', webpage, 'title')
        auth = self._html_search_regex(r'[^>]+.authType = "(.+?)"', webpage, 'authType')
        if "unauth" in auth:
            auth_required = 'false'
        if "auth" in auth:
            auth_required = 'true'
        videoType = self._html_search_regex(r'[^>]+.videoType = "(.+?)"', webpage, 'videoType')
````

**Comment by JamesTheAwesomeDude** [comment]
Trying to trigger that code you added, but `--ap-mso=unauth` didn't work

**Comment by TheDaChicken** [comment]
Ok! I will try again, The annoying thing is that youtube-dl stored my TV Provider were even if I don't use it. It uses it. So, I guess, I am going to have to play around with the code and try again.

**Comment by TheDaChicken** [comment]
Ok! Now try it again. Sorry, about that. lol

**Comment by JamesTheAwesomeDude** [comment]
0.o apparently that latest patch didn't fix it, either

````
[CartoonNetwork] when-it-rains: Downloading webpage
[CartoonNetwork] 0e9df1345ed47078df3cf75bec73f74562a4fdb8: Downloading JSON metadata
ERROR: This video is only available for users of participating TV providers. Use --ap-mso to specify Adobe Pass Multiple-system operator Identifier and --ap-username and --ap-password or --netrc to provide account credentials.
Traceback (most recent call last):
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/YoutubeDL.py", line 792, in extract_info
    ie_result = ie.extract(url)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/common.py", line 502, in extract
    ie_result = self._real_extract(url)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/cartoonnetwork.py", line 46, in _real_extract
    'auth_required': auth_required,
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/turner.py", line 212, in _extract_ngtv_info
    m3u8_url, media_id, ap_data or {}, tokenizer_query)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/turner.py", line 40, in _add_akamai_spe_token
    query['accessToken'] = self._extract_mvpd_auth(ap_data['url'], content_id, ap_data['site_name'], ap_data['site_name'])
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/adobepass.py", line 1409, in _extract_mvpd_auth
    raise_mvpd_required()
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/adobepass.py", line 1374, in raise_mvpd_required
    'and --ap-username and --ap-password or --netrc to provide account credentials.', expected=True)
youtube_dl.utils.ExtractorError: This video is only available for users of participating TV providers. Use --ap-mso to specify Adobe Pass Multiple-system operator Identifier and --ap-username and --ap-password or --netrc to provide account credentials.
````

But it's working on your end, without TV provider login?

**Comment by TheDaChicken** [comment]
Ya, it's working on my end without TV Provider login. Weird. I am going to try another way and see if that works.

**Comment by JamesTheAwesomeDude** [comment]
>Now try uninstall youtube-dl and then try installing from the fork again.

* `python3 -m pip uninstall youtube-dl`
* `git pull` to dfd8f364728c47ed514e32deb739a264bc5f2966
* `python3 -m pip install .`

````
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['https://www.cartoonnetwork.com/video/unlocked/when-it-rains-episode.html', '-v', '--hls-prefer-native']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.08.04
[debug] Python version 3.5.2 (CPython) - Linux-4.10.0-38-generic-x86_64-with-Ubuntu-16.04-xenial
[debug] exe versions: ffmpeg 3.4.1-1, ffprobe 3.4.1-1
[debug] Proxy map: {}
[CartoonNetwork] when-it-rains: Downloading webpage
[CartoonNetwork] 0e9df1345ed47078df3cf75bec73f74562a4fdb8: Downloading JSON metadata
ERROR: This video is only available for users of participating TV providers. Use --ap-mso to specify Adobe Pass Multiple-system operator Identifier and --ap-username and --ap-password or --netrc to provide account credentials.
Traceback (most recent call last):
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/YoutubeDL.py", line 792, in extract_info
    ie_result = ie.extract(url)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/common.py", line 502, in extract
    ie_result = self._real_extract(url)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/cartoonnetwork.py", line 43, in _real_extract
    'auth_required': 'true',
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/turner.py", line 212, in _extract_ngtv_info
    m3u8_url, media_id, ap_data or {}, tokenizer_query)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/turner.py", line 40, in _add_akamai_spe_token
    query['accessToken'] = self._extract_mvpd_auth(ap_data['url'], content_id, ap_data['site_name'], ap_data['site_name'])
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/adobepass.py", line 1409, in _extract_mvpd_auth
    raise_mvpd_required()
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/adobepass.py", line 1374, in raise_mvpd_required
    'and --ap-username and --ap-password or --netrc to provide account credentials.', expected=True)
youtube_dl.utils.ExtractorError: This video is only available for users of participating TV providers. Use --ap-mso to specify Adobe Pass Multiple-system operator Identifier and --ap-username and --ap-password or --netrc to provide account credentials.

````

I swear I'm not just trying to be difficult I don't know why it's not working

![crying cat meme image 240p thumbnail](https://i.imgur.com/q4cvXyrm.jpg)

**Comment by TheDaChicken** [comment]
Ok! Try it now..

**Comment by JamesTheAwesomeDude** [comment]
Running c17af258168bc75dc283265c675f286256ff7a0c:

````
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-v', '--hls-prefer-native', 'https://www.cartoonnetwork.com/video/unlocked/when-it-rains-episode.html', '--write-pages']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.08.04
[debug] Python version 3.5.2 (CPython) - Linux-4.10.0-38-generic-x86_64-with-Ubuntu-16.04-xenial
[debug] exe versions: ffmpeg 3.4.1-1, ffprobe 3.4.1-1
[debug] Proxy map: {}
[CartoonNetwork] when-it-rains: Downloading webpage
[CartoonNetwork] Saving request to when-it-rains_https_-_www.cartoonnetwork.com_video_unlocked_when-it-rains-episode.html.dump
[CartoonNetwork] 0e9df1345ed47078df3cf75bec73f74562a4fdb8: Downloading JSON metadata
[CartoonNetwork] Saving request to 0e9df1345ed47078df3cf75bec73f74562a4fdb8_http_-_medium.ngtv.io_media_0e9df1345ed47078df3cf75bec73f74562a4fdb8_tv.dump
ERROR: This video is only available for users of participating TV providers. Use --ap-mso to specify Adobe Pass Multiple-system operator Identifier and --ap-username and --ap-password or --netrc to provide account credentials.
Traceback (most recent call last):
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/YoutubeDL.py", line 792, in extract_info
    ie_result = ie.extract(url)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/common.py", line 502, in extract
    ie_result = self._real_extract(url)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/cartoonnetwork.py", line 46, in _real_extract
    'auth_required': auth_required,
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/turner.py", line 212, in _extract_ngtv_info
    m3u8_url, media_id, ap_data or {}, tokenizer_query)
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/turner.py", line 40, in _add_akamai_spe_token
    query['accessToken'] = self._extract_mvpd_auth(ap_data['url'], content_id, ap_data['site_name'], ap_data['site_name'])
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/adobepass.py", line 1409, in _extract_mvpd_auth
    raise_mvpd_required()
  File "/home/james/.local/lib/python3.5/site-packages/youtube_dl/extractor/adobepass.py", line 1374, in raise_mvpd_required
    'and --ap-username and --ap-password or --netrc to provide account credentials.', expected=True)
youtube_dl.utils.ExtractorError: This video is only available for users of participating TV providers. Use --ap-mso to specify Adobe Pass Multiple-system operator Identifier and --ap-username and --ap-password or --netrc to provide account credentials.

````

* [`0e9df1345ed47078df3cf75bec73f74562a4fdb8_http_-_medium.ngtv.io_media_0e9df1345ed47078df3cf75bec73f74562a4fdb8_tv.dump`](http://ix.io/1m08)
* [`when-it-rains_https_-_www.cartoonnetwork.com_video_unlocked_when-it-rains-episode.html.dump`](http://ix.io/1m9b)

**Comment by JamesTheAwesomeDude** [comment]
@TheDaChicken you're awesome btw. Full deets in the other thread but tl;dr is you did it (at least now it Works On _My_ Machine™, too!)

**Comment by TheDaChicken** [comment]
Thanks!

**Comment by tgray** [comment]
Was this ever fixed?

**Comment by eduardog3000** [comment]
@tgray #17271 is a PR for a branch where this was fixed. It hasn't been merged since it seems the main collaborator doesn't look at a PR more than once. If you want to use CN you'll have to pull that branch.

**Comment by Danielperalta07** [comment]
@dstftw could you please sir take a look at this? Much appreciated!

**Comment by JamesTheAwesomeDude** [comment]
@tgray this is not fixed yet…@TheDaChicken wrote a patch (#17271), but it's not being accepted by @dstftw for some reason.

I went onto the PR thread to try to beg for some clarification, but it's not going to get a productive response; here's my (SPECULATIVE) attempt at interpreting the back-and-forth from that bizarre thread:

* #17271 fixes the bug, but contains _some kind of_ technical deficiency making it insuitable for merge with the master codebase.
* @dstftw feels he's fully specified the technical deficiency, and is simply waiting for us to stop asking him to handfeed us, and just FIX THE DANG PR so he can merge it.
* …however, he didn't actually provide enough information for @TheDaChicken (or anyone else) to fix whatever it is. So we've reached a deadlock where every party thinks the other is being intentionally obtuse.
* (EXTRA-SPECULATIVE:) The technical deficiency actually comprises _the introduction of (??) a new bug_, one which would require someone with a Cartoon Network premium account to even begin to fix, so leaving the entire `CartoonNetwork`handler broken for someone to independently fix (along with its auth code) later on down the road will actually result in less wasted coder-hours in the long run.
